control-plane-operator/controllers/awsprivatelink/awsprivatelink_controller.go

```diff
@@ -187,12 +189,113 @@ const (
 // the existence of AWS Endpoints for it in the guest cluster infrastructure.
 type AWSEndpointServiceReconciler struct {
 	client.Client
-	ec2Client     ec2iface.EC2API
-	route53Client route53iface.Route53API
 	upsert.CreateOrUpdateProvider
-	AssumeEndpointRoleARN *string
-	AssumeRoute53RoleARN  *string
-	LocalZoneID           *string
+	awsClientBuilder clientBuilder
+}
+
+type clientBuilder struct {
+	mu                    sync.Mutex
+	initialized           bool
+	assumeEndpointRoleARN string
+	assumeRoute53RoleARN  string
+	localZoneID           string
+}
+
+func (b *clientBuilder) awsSession() (*session.Session, error) {
+	s, err := session.NewSession()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create AWS session: %w", err)
+	}
+	s.Handlers.Build.PushBackNamed(request.NamedHandler{
+		Name: "openshift.io/hypershift",
+		Fn:   request.MakeAddToUserAgentHandler("openshift.io hypershift", "control-plane-operator"),
+	})
+	return s, nil
+}
+
+func (b *clientBuilder) getClients() (ec2iface.EC2API, route53iface.Route53API, error) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	if !b.initialized {
+		return nil, nil, errors.New("clients not initialized")
+	}
+
+	// AWS_SHARED_CREDENTIALS_FILE and AWS_REGION envvar should be set in operator deployment
+	awsEndpointSession, err := b.awsSession()
+	if err != nil {
+		return nil, nil, err
+	}
+	if b.assumeEndpointRoleARN != "" {
+		awsEndpointSession.Config.WithCredentials(stscreds.NewCredentials(awsEndpointSession, b.assumeEndpointRoleARN))
+	}
+	awsRoute53Session, err := b.awsSession()
+	if err != nil {
+		return nil, nil, err
+	}
+	if b.assumeRoute53RoleARN != "" {
+		awsRoute53Session.Config.WithCredentials(stscreds.NewCredentials(awsRoute53Session, b.assumeRoute53RoleARN))
+	}
+	awsConfig := aws.NewConfig()
+	ec2Client := ec2.New(awsEndpointSession, awsConfig)
+	route53Config := aws.NewConfig()
+	// Hardcode region for route53 config
+	route53Config.Region = aws.String("us-east-1")
+	route53Client := route53.New(awsRoute53Session, route53Config)
+
+	return ec2Client, route53Client, nil
+}
+
+func (b *clientBuilder) localHostedZoneID() string {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	return b.localZoneID
+}
+
+func (b *clientBuilder) initializeWithHCP(log logr.Logger, hcp *hyperv1.HostedControlPlane) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	if !b.initialized {
```

Comment:
would it make any difference if this func is just
```
b.warnOnDifferentValues(log, hcp)
b.setFromHCP(hcp)
b.initialized = true
```
without the conditional?

Reply:
we'd warn unnecessarily when initially setting the values